src/App.tsx

```diff
@@ -1,14 +1,69 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, { useState, useEffect } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+import { useAppSelector } from './app/hooks';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[] | null>(null);
+  const [visibleTodos, setVisibleTodos] = useState<Todo[] | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const selectedTodo = useAppSelector(state => state.currentTodo);
+  const query = useAppSelector(state => state.filter.query);
+  const status = useAppSelector(state => state.filter.status);
+
+  useEffect(() => {
+    const loadTodos = async () => {
+      setIsLoading(true);
+      const response = await getTodos();
+
+      setTodos(response);
+      setIsLoading(false);
+    };
+
+    loadTodos();
+  }, []);
+
+  useEffect(() => {
+    if (!todos) {
+      return;
+    }
+
+    let processedTodos = [...todos];
+
+    if (query) {
+      processedTodos = processedTodos.filter(todo => (
+        todo.title.toLocaleLowerCase().includes(query.toLocaleLowerCase())
+      ));
+    }
+
+    processedTodos = processedTodos.filter(todo => {
+      switch (status) {
+        case 'active':
+          return !todo.completed;
+        case 'completed':
+          return todo.completed;
+        default:
+          return todo;
+      }
+    });
+
+    if (!processedTodos.length) {
+      setVisibleTodos(null);
+
+      return;
+    }
+
+    setVisibleTodos(processedTodos);
+  }, [todos, query, status]);
```

Comment:
I don't think that you need to use a useEffect here and this processedTodos variable. Can you just move visibleTodos to a variable with useMemo()? What do you think about it?

Reply:
done